src/Base/Exception.h

```diff
@@ -33,6 +33,13 @@
 #include "FileInfo.h"
 #include "BaseClass.h"
 
+/* MACROS FOR THROWING EXCEPTIONS */
+
+#define THROW(exception) throw Base::exception(__FILE__,__LINE__,__PRETTY_FUNCTION__);
+#define THROWM(exception, message) throw Base::exception(message,__FILE__,__LINE__,__PRETTY_FUNCTION__);

```

Comment:
> It is intended.

The only problem is that third party libraries may use the same class names (e.g. PyCXX). Then you get an ambiguity error and can't resolve it. Better to force to use namespace + class name when using the macro.

> As for typing, I think it is one of those corners where c++ forces you to copy/paste, as you have to define the constructors for each class. 

Instead of making too many overloaded constructors and repeating this for any sub-class one could write setter methods in the base class. This reduces the effort dramatically.